File: server/DaysUntil.Tests/Unit/Controller/AuthControllerTest.cs
using Moq;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging.Abstractions;
using Daysuntil.Controllers;
using Daysuntil.Dtos;

public class AuthControllerTest
{
    private readonly Mock<IUserService> _userServiceMock;
    private readonly AuthController _controller;

    public AuthControllerTest()
    {
        _userServiceMock = new Mock<IUserService>();
        _controller = new AuthController(_userServiceMock.Object, new NullLogger<AuthController>());
    }

    [Fact]
    public async Task TestRegisterSuccess()
    {
        var data = new RegisterDto { Username = "testuser", Password = "testpass", Email = "test@test.com" };
        var token = "example_token";
        _userServiceMock.Setup(x => x.RegisterAsync(data.Username, data.Email, data.Password))
            .ReturnsAsync(token);

        var result = await _controller.Register(data) as OkObjectResult;

        Assert.NotNull(result);
        Assert.Equal(200, result.StatusCode);
        Assert.Equal(token, result.Value);
    }

    [Fact]
    public async Task TestRegisterFailureMissingFields()
    {
        var data = new RegisterDto { Username = "testuser", Email = "", Password = "" };
        _controller.ModelState.AddModelError("Password", "Required");

        var result = await _controller.Register(data) as BadRequestObjectResult;

        Assert.NotNull(result);
        Assert.Equal(400, result.StatusCode);
    }

    [Fact]
    public async Task TestLoginSuccess()
    {
        var data = new LoginDto { Username = "testuser", Password = "validPassword" };
        var token = "example_token";
        _userServiceMock.Setup(x => x.LoginAsync(data.Username, data.Password))
            .ReturnsAsync(token);

        var result = await _controller.Login(data) as OkObjectResult;

        Assert.NotNull(result);
        Assert.Equal(200, result.StatusCode);

        Assert.Equal(token, result.Value);
    }

    [Fact]
    public async Task TestLoginFailure()
    {
        var data = new LoginDto { Username = "testuser", Password = "invalidPassword" };
        _userServiceMock.Setup(x => x.LoginAsync(data.Username, data.Password))
            .ReturnsAsync((string)null);

        var result = await _controller.Login(data);
        Assert.IsType<UnauthorizedResult>(result);
    }
}

File: server/php/tests/Unit/Controller/AuthControllerTest.php
<?php

use PHPUnit\Framework\TestCase;
use DaysUntil\Controller\AuthController;
use DaysUntil\Service\UserService;
use Firebase\JWT\JWT;

class AuthControllerTest extends TestCase
{
    private $controller;
    private $userService;

    protected function setUp(): void
    {
        $this->userService = $this->createMock(UserService::class);
        $this->controller = new AuthController($this->userService);
    }

    public function testRegisterSuccess()
    {
        $data = ['username' => 'testuser', 'password' => 'testpass', 'email' => 'test@test.com'];
        $userId = 1;

        $this->userService->method('register')
            ->with($data['username'], $data['email'], $data['password'])
            ->willReturn($userId);

        $this->userService->expects($this->once())
            ->method('saveToken')
            ->with($this->equalTo($userId), $this->anything());

        ob_start();
        $this->controller->register($data);
        $output = json_decode(ob_get_clean(), true);

        $this->assertEquals(201, http_response_code());
        $this->assertArrayHasKey('token', $output);
        $this->assertEquals('User successfully registered', $output['message']);
        $this->assertEquals($userId, $output['userId']);
    }

    public function testRegisterFailureMissingFields()
    {
        $data = ['username' => 'testuser'];

        ob_start();
        $this->controller->register($data);
        $output = json_decode(ob_get_clean(), true);

        $this->assertEquals(400, http_response_code());
        $this->assertEquals('Missing required fields', $output['message']);
    }
    
    public function testLoginSuccess()
    {
        $data = ['username' => 'testuser', 'password' => 'validPassword'];
        $user = ['id' => 1];
    
        $this->userService->method('login')
            ->willReturn($user);
    
        $this->userService->expects($this->once())
            ->method('saveToken')
            ->with($this->equalTo($user['id']), $this->anything());
    
        ob_start();
        $this->controller->login($data);
        $output = json_decode(ob_get_clean(), true);
    
        $this->assertEquals(200, http_response_code());
        $this->assertArrayHasKey('token', $output);
        $this->assertEquals('Login successful', $output['message']);
    }

    public function testLoginFailure()
    {
        $data = ['username' => 'testuser', 'password' => 'invalidPassword'];

        $this->userService->method('login')
            ->with($data['username'], $data['password'])
            ->willReturn(null);

        ob_start();
        $this->controller->login($data);
        $output = json_decode(ob_get_clean(), true);

        $this->assertEquals(401, http_response_code());
        $this->assertEquals('Login failed', $output['message']);
    }

    public function testValidateTokenSuccess()
    {
        $userId = 1;
        $token = JWT::encode(['sub' => $userId], 'my_super_secret_key', 'HS256');
        $headers = ['Authorization' => "Bearer $token"];

        $result = $this->controller->validateToken($headers);

        $this->assertEquals($userId, $result['sub']);
    }

    public function testValidateTokenFailure()
    {
        $headers = ['Authorization' => "Bearer invalid_token"];
        $this->expectException(Exception::class);
        $this->expectExceptionMessage("Invalid token:");

        $this->controller->validateToken($headers);
    }
}
